### ISSUE TITLE
DNS cannot route to direct outbound

### ISSUE THREAD

**Issue by reeceyng** [SagerNet/sing-box#7]
### Welcome

- [X] Yes, I'm using the latest major release. Only such installations are supported.
- [X] Yes, I've searched similar issues on GitHub and didn't find any.
- [X] Yes, I've included all information below (version, config, etc).

### Description of the problem

I added three domain names to the direct routing, including route and dns route.

But, the speed of these websites was still very slow. And I found it's using alidns from remote service to resolve these 3 website.

However the connection was direct. So I think it's only my dns route fail.

Relevant logs:
<details>

```
DEBUG[0631] [3584963253] router: sniffed protocol: tls, domain: www.bilibili.com
DEBUG[0631] [3584963253] router: match[1] domain_keyword=[bilibili.com bilivideo.com netease.com] => direct-out
INFO[0631] [3584963253] outbound/direct[direct-out]: outbound connection to [::ffff:45.40.60.170]:443
INFO[0633] [127784800] inbound/tproxy[tproxy-in]: inbound connection to [::ffff:45.40.60.170]:443
DEBUG[0633] [127784800] router: sniffed protocol: tls, domain: api.vc.bilibili.com
DEBUG[0633] [127784800] router: match[1] domain_keyword=[bilibili.com bilivideo.com netease.com] => direct-out
INFO[0633] [127784800] outbound/direct[direct-out]: outbound connection to [::ffff:45.40.60.170]:443
```

</details>

### Version of sing-box

<details>
❯ ./sing-box version

sing-box version nightly (go1.19 linux/amd64)
</details>

### Server and client configuration file

Server configuration
<details>

```
{
  "inbounds": [
    {
      "type": "shadowsocks",
      "listen": "::",
      "listen_port": 8080,
      "method": "2022-blake3-aes-128-gcm",
      "password": "8JCsPssfgS8tiRwiMlhARg=="
    }
  ]
}
```

</details>

Client configuration
<details>

```
{
  "dns": {
    "servers": [
      {
        "tag": "google",
        "address": "8.8.8.8"
      },
      {
        "tag": "alidns",
        "address": "223.5.5.5",
        "address_strategy": "prefer_ipv4"
      }
    ],
    "rules": [
      {
        "domain_keyword": [
          "bilibili.com",
          "bilivideo.com",
          "netease.com"
          ],
        "invert": false,
        "server": "alidns"
      }
    ]
  },
  "inbounds": [
    {
      "type": "tproxy",
      "tag": "tproxy-in",

      "listen": "::",
      "listen_port": 10800,
      "sniff": true
    }
  ],
  "outbounds": [
    {
      "type": "shadowsocks",
      "server": "my.domain",
      "server_port": 8080,
      "method": "2022-blake3-aes-128-gcm",
      "password": "8JCsPssfgS8tiRwiMlhARg=="
    },
    {
      "type": "direct",
      "tag": "direct-out"
    },
    {
      "type": "dns",
      "tag": "dns-out"
    }
  ],
  "route": {
    "rules": [
      {
        "port": 53,
        "outbound": "dns-out"
      },
      {
        "domain_keyword": [
          "bilibili.com",
          "bilivideo.com",
          "netease.com"
          ],
        "invert": false,
        "outbound": "direct-out"
      },
      {
        "ip_cidr": [
          "223.5.5.5/32"
        ],
        "invert": false,
        "outbound": "direct-out"
      }
    ],
    "auto_detect_interface": true
  }
}

```

</details>

**Comment by nekohasekai** [comment]
```diff
     {
        "tag": "alidns",
        "address": "223.5.5.5",
        "address_strategy": "prefer_ipv4"
+       "detour": "direct-out"
      }
```

**Comment by nekohasekai** [comment]
For https://github.com/SagerNet/sing-box/issues/4#issuecomment-1207483803:

Large iperf3 packet testing (> 16k) is not supported, and the panic has been fixed in https://github.com/SagerNet/sing-box/commit/537f81ebbaa4a1ce2ce574106e64456abd782d94

**Comment by reeceyng** [comment]
> ```diff
>      {
>         "tag": "alidns",
>         "address": "223.5.5.5",
>         "address_strategy": "prefer_ipv4"
> +       "detour": "direct-out"
>       }
> ```

Thanks! I thoungt it was only effect to tun mode. It works now!

**Comment by reeceyng** [comment]
> For [#4 (comment)](https://github.com/SagerNet/sing-box/issues/4#issuecomment-1207483803):
> 
> iperf3 large packet UDP testing (> 16k) is not supported, and the panic has been fixed in [2c6d239](https://github.com/SagerNet/sing-box/commit/2c6d23952518cc633fb219b29f49de6c6182c48e)

I found smux with udp dns doesn't work and dot or doh is fine.

**Comment by nekohasekai** [comment]
> > For [#4 (comment)](https://github.com/SagerNet/sing-box/issues/4#issuecomment-1207483803):
> > iperf3 large packet UDP testing (> 16k) is not supported, and the panic has been fixed in [2c6d239](https://github.com/SagerNet/sing-box/commit/2c6d23952518cc633fb219b29f49de6c6182c48e)
> 
> I found smux with udp dns doesn't work and dot or doh is fine.

Create an new issue.
